File: examples/spool-writer/typescript/spool-writer.mjs
// A spool writer without the SDK (protocol/spool-format.md, D66): what a team
// instrumenting a provider SDK themselves needs — minute windows per
// (tag, versionId, arm, model, status, errorClass), the fixed latency
// buckets, segment naming, `.open` → fsync → rename, and rotation on the
// minute or at 1 MiB. `airprompterd` uploads what lands in `dir`. No
// dependencies; `conformance/run.mjs` drives it through vectors/spool.json.
//
//   const spool = new SpoolWriter({ dir: `${stateDir}/airprompter/${agentId}/${target}/spool/telemetry`, instanceId, sdk: "acme-logger/1.0" });
//   spool.observe({ tag, versionId, arm, model, status: "ok", latencyMs: 812, tokens: { input: 400, output: 90 } });
//   spool.feedback({ tag, versionId, arm, model, outcomes: { accepted: true } });
//   process.on("beforeExit", () => spool.close());

import { closeSync, fsyncSync, mkdirSync, openSync, renameSync, writeSync } from "node:fs";
import { join } from "node:path";

export const LATENCY_BUCKET_EDGES_MS = [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 65536];
export const SEGMENT_MAX_BYTES = 1024 * 1024;
const OUTCOME_NAME = /^[a-z][a-zA-Z0-9]{0,31}$/;

export const latencyBucketIndex = (ms) => {
  const index = LATENCY_BUCKET_EDGES_MS.findIndex((edge) => ms <= edge);
  return index === -1 ? LATENCY_BUCKET_EDGES_MS.length - 1 : index;
};
export const epochMinute = (epochMs) => Math.floor(epochMs / 60000);
export const minuteOf = (epochMs) => new Date(epochMinute(epochMs) * 60000).toISOString().replace(".000Z", "Z");
export const segmentName = (instanceId, minute, n) => `seg-${instanceId}-${minute}-${n}.ndjson`;

export class SpoolWriter {
  constructor({ dir = null, instanceId, instanceClass = "resident", sdk = "spool-writer-example/1.0", now = Date.now }) {
    Object.assign(this, { dir, instanceId, instanceClass, sdk, now, open: new Map(), openMinute: null, segMinute: null, segBytes: 0, segN: -1, sealed: false, emitted: [] });
    if (dir) mkdirSync(dir, { recursive: true, mode: 0o700 });
  }

  /** One model call: latency, tokens, status and error class — never text, ids or messages. */
  observe({ tag, versionId, arm = "none", model, status = "ok", errorClass = null, usageSource = "reported", latencyMs, tokens = {}, checks = null, outcomes = null }, at = this.now()) {
    const row = this.window(at, { tag, versionId, arm, model, status, errorClass, usageSource });
    row.count += 1;
    row.latencyMs.buckets[latencyBucketIndex(latencyMs)] += 1;
    row.latencyMs.sum += Math.max(0, Math.round(latencyMs));
    row.tokens.input += tokens.input ?? 0;
    row.tokens.output += tokens.output ?? 0;
    if (tokens.cachedInput) row.tokens.cachedInput = (row.tokens.cachedInput ?? 0) + tokens.cachedInput;
    if (checks) row.checks = { passed: (row.checks?.passed ?? 0) + (checks.passed ?? 0), failed: (row.checks?.failed ?? 0) + (checks.failed ?? 0) };
    if (outcomes) mergeOutcomes(row, outcomes);
  }

  /** Feedback filed against a run: rides on the run's `ok` window for this minute, never adds to count or latency. */
  feedback({ tag, versionId, arm = "none", model, outcomes }, at = this.now()) {
    mergeOutcomes(this.window(at, { tag, versionId, arm, model, status: "ok", errorClass: null, usageSource: "reported" }), outcomes);
  }

  window(at, d) {
    const minute = minuteOf(at);
    if (this.openMinute !== null && this.openMinute !== minute) this.close(at);
    this.openMinute = minute;
    const key = JSON.stringify([d.tag, d.versionId, d.arm, d.model, d.status, d.errorClass ?? null]);
    let row = this.open.get(key);
    if (!row) {
      row = { type: "window", v: 1, minute, instanceId: this.instanceId, instanceClass: this.instanceClass, ...d, errorClass: d.errorClass ?? null, count: 0, latencyMs: { buckets: new Array(16).fill(0), sum: 0 }, tokens: { input: 0, output: 0 }, sdk: this.sdk };
      this.open.set(key, row);
    }
    return row;
  }

  /** Writes the open minute's rows to a segment (`.open`, fsync, rename) and starts a new minute. */
  close(at = this.now()) {
    const rows = [...this.open.values()];
    this.open.clear();
    this.openMinute = null;
    for (const row of rows) this.append(JSON.stringify(row) + "\n", at);
    this.emitted.push(...rows);
    if (this.fd !== undefined) this.sealSegment();
  }

  /** Which segment a line lands in: a new one on a new minute or when this line would push past 1 MiB. */
  planSegment(epochMs, lineBytes) {
    const minute = epochMinute(epochMs);
    const rotated = this.sealed || this.segMinute !== minute || this.segBytes + lineBytes > SEGMENT_MAX_BYTES;
    this.sealed = false;
    if (rotated) {
      this.segN = this.segMinute === minute ? this.segN + 1 : 0;
      this.segMinute = minute;
      this.segBytes = 0;
    }
    this.segBytes += lineBytes;
    return { segment: segmentName(this.instanceId, this.segMinute, this.segN), rotated };
  }

  append(line, at) {
    const bytes = Buffer.byteLength(line, "utf8");
    const { segment, rotated } = this.planSegment(at, bytes);
    if (!this.dir) return;
    if (rotated && this.fd !== undefined) this.sealSegment();
    if (this.fd === undefined) (this.segPath = join(this.dir, segment)), (this.fd = openSync(this.segPath + ".open", "a", 0o600));
    writeSync(this.fd, line);
  }

  sealSegment() {
    fsyncSync(this.fd);
    closeSync(this.fd);
    renameSync(this.segPath + ".open", this.segPath);
    this.fd = undefined;
    this.sealed = true; // a later line in the same minute opens the next segment, never this sealed one
  }
}

function mergeOutcomes(row, outcomes) {
  row.outcomes ??= {};
  for (const [name, value] of Object.entries(outcomes)) {
    const numeric = typeof value === "boolean" ? (value ? 1 : 0) : value;
    if (!OUTCOME_NAME.test(name) || typeof numeric !== "number" || !Number.isFinite(numeric)) continue;
    const current = row.outcomes[name] ?? { n: 0, sum: 0 };
    row.outcomes[name] = { n: current.n + 1, sum: current.sum + numeric };
  }
}

File: examples/spool-writer/python/spool_writer.py
"""A spool writer without the SDK (protocol/spool-format.md, D66): what a team
instrumenting a provider SDK themselves needs — minute windows per
(tag, versionId, arm, model, status, errorClass), the fixed latency buckets,
segment naming, ``.open`` → fsync → rename, and rotation on the minute or at
1 MiB. ``airprompterd`` uploads what lands in ``dir``. Standard library only;
``check_vectors.py`` drives it through protocol/vectors/spool.json.

::

    spool = SpoolWriter(dir=f"{state_dir}/airprompter/{agent_id}/{target}/spool/telemetry", instance_id=instance_id, sdk="acme-logger/1.0")
    spool.observe(tag=tag, version_id=version_id, arm=arm, model=model, status="ok", latency_ms=812, tokens={"input": 400, "output": 90})
    spool.feedback(tag=tag, version_id=version_id, arm=arm, model=model, outcomes={"accepted": True})
    atexit.register(spool.close)
"""

from __future__ import annotations

import json
import os
import re
import time
from datetime import datetime, timezone
from typing import Any, Mapping, Optional

LATENCY_BUCKET_EDGES_MS = [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 65536]
SEGMENT_MAX_BYTES = 1024 * 1024
_OUTCOME_NAME = re.compile(r"^[a-z][a-zA-Z0-9]{0,31}$")


def latency_bucket_index(ms: float) -> int:
    return next((i for i, edge in enumerate(LATENCY_BUCKET_EDGES_MS) if ms <= edge), len(LATENCY_BUCKET_EDGES_MS) - 1)


def epoch_minute(epoch_ms: float) -> int:
    return int(epoch_ms // 60000)


def minute_of(epoch_ms: float) -> str:
    return datetime.fromtimestamp(epoch_minute(epoch_ms) * 60, tz=timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def segment_name(instance_id: str, minute: int, n: int) -> str:
    return f"seg-{instance_id}-{minute}-{n}.ndjson"


class SpoolWriter:
    def __init__(self, *, dir: Optional[str] = None, instance_id: str, instance_class: str = "resident", sdk: str = "spool-writer-example/1.0", now=lambda: time.time() * 1000):  # noqa: A002
        self.dir, self.instance_id, self.instance_class, self.sdk, self.now = dir, instance_id, instance_class, sdk, now
        self.open: dict[str, dict[str, Any]] = {}
        self.open_minute: Optional[str] = None
        self.seg_minute: Optional[int] = None
        self.seg_bytes, self.seg_n, self.sealed = 0, -1, False
        self.fd: Optional[int] = None
        self.seg_path = ""
        self.emitted: list[dict[str, Any]] = []
        if dir:
            os.makedirs(dir, mode=0o700, exist_ok=True)

    def observe(self, *, tag: str, version_id: str, model: str, latency_ms: float, arm: str = "none", status: str = "ok", error_class: Optional[str] = None, usage_source: str = "reported", tokens: Optional[Mapping[str, int]] = None, checks: Optional[Mapping[str, int]] = None, outcomes: Optional[Mapping[str, Any]] = None, at: Optional[float] = None) -> None:
        """One model call: latency, tokens, status and error class — never text, ids or messages."""
        row = self._window(self.now() if at is None else at, tag, version_id, arm, model, status, error_class, usage_source)
        row["count"] += 1
        row["latencyMs"]["buckets"][latency_bucket_index(latency_ms)] += 1
        row["latencyMs"]["sum"] += max(0, int(latency_ms + 0.5))
        tokens = tokens or {}
        row["tokens"]["input"] += tokens.get("input", 0)
        row["tokens"]["output"] += tokens.get("output", 0)
        if tokens.get("cachedInput"):
            row["tokens"]["cachedInput"] = row["tokens"].get("cachedInput", 0) + tokens["cachedInput"]
        if checks:
            previous = row.get("checks") or {"passed": 0, "failed": 0}
            row["checks"] = {"passed": previous["passed"] + checks.get("passed", 0), "failed": previous["failed"] + checks.get("failed", 0)}
        if outcomes:
            _merge_outcomes(row, outcomes)

    def feedback(self, *, tag: str, version_id: str, model: str, outcomes: Mapping[str, Any], arm: str = "none", at: Optional[float] = None) -> None:
        """Feedback filed against a run: rides on the run's ``ok`` window for this minute, never adds to count or latency."""
        _merge_outcomes(self._window(self.now() if at is None else at, tag, version_id, arm, model, "ok", None, "reported"), outcomes)

    def _window(self, at: float, tag: str, version_id: str, arm: str, model: str, status: str, error_class: Optional[str], usage_source: str) -> dict[str, Any]:
        minute = minute_of(at)
        if self.open_minute is not None and self.open_minute != minute:
            self.close(at)
        self.open_minute = minute
        key = json.dumps([tag, version_id, arm, model, status, error_class])
        row = self.open.get(key)
        if row is None:
            row = {"type": "window", "v": 1, "minute": minute, "instanceId": self.instance_id, "instanceClass": self.instance_class, "tag": tag, "versionId": version_id, "arm": arm, "model": model, "status": status, "errorClass": error_class, "usageSource": usage_source, "count": 0, "latencyMs": {"buckets": [0] * 16, "sum": 0}, "tokens": {"input": 0, "output": 0}, "sdk": self.sdk}
            self.open[key] = row
        return row

    def close(self, at: Optional[float] = None) -> None:
        """Writes the open minute's rows to a segment (``.open``, fsync, rename) and starts a new minute."""
        at = self.now() if at is None else at
        rows = list(self.open.values())
        self.open.clear()
        self.open_minute = None
        for row in rows:
            self._append(json.dumps(row, separators=(",", ":")) + "\n", at)
        self.emitted.extend(rows)
        if self.fd is not None:
            self._seal()

    def plan_segment(self, epoch_ms: float, line_bytes: int) -> tuple[str, bool]:
        """Which segment a line lands in: a new one on a new minute or when this line would push past 1 MiB."""
        minute = epoch_minute(epoch_ms)
        rotated = self.sealed or self.seg_minute != minute or self.seg_bytes + line_bytes > SEGMENT_MAX_BYTES
        self.sealed = False
        if rotated:
            self.seg_n = self.seg_n + 1 if self.seg_minute == minute else 0
            self.seg_minute, self.seg_bytes = minute, 0
        self.seg_bytes += line_bytes
        return segment_name(self.instance_id, self.seg_minute, self.seg_n), rotated

    def _append(self, line: str, at: float) -> None:
        data = line.encode("utf-8")
        segment, rotated = self.plan_segment(at, len(data))
        if not self.dir:
            return
        if rotated and self.fd is not None:
            self._seal()
        if self.fd is None:
            self.seg_path = os.path.join(self.dir, segment)
            self.fd = os.open(self.seg_path + ".open", os.O_WRONLY | os.O_CREAT | os.O_APPEND, 0o600)
        os.write(self.fd, data)

    def _seal(self) -> None:
        assert self.fd is not None
        os.fsync(self.fd)
        os.close(self.fd)
        os.rename(self.seg_path + ".open", self.seg_path)
        self.fd = None
        self.sealed = True  # a later line in the same minute opens the next segment, never this sealed one


def _merge_outcomes(row: dict[str, Any], outcomes: Mapping[str, Any]) -> None:
    merged = row.setdefault("outcomes", {})
    for name, value in outcomes.items():
        numeric = (1 if value else 0) if isinstance(value, bool) else value
        if not _OUTCOME_NAME.match(name) or not isinstance(numeric, (int, float)) or numeric != numeric or numeric in (float("inf"), float("-inf")):
            continue
        current = merged.get(name, {"n": 0, "sum": 0})
        merged[name] = {"n": current["n"] + 1, "sum": current["sum"] + numeric}
